File: src/commands/sale.rs
use crate::chain::client::{provider, signer_client};
use crate::chain::contracts::{
    AgentRaiseFactoryContract, Erc20Metadata, SaleContract, read_factory_project,
    read_sale_collateral, read_sale_project_id, token_meta,
};
use crate::cli::{Cli, SaleCommand, SaleTarget};
use crate::config::RuntimeConfig;
use crate::output::{format_token_amount, print_receipt};
use crate::util::{parse_amount_raw, parse_amount_units, send_and_wait};
use anyhow::{Result, bail};
use ethers::prelude::*;
use std::sync::Arc;

pub async fn run_sale(cli: &Cli, cfg: &RuntimeConfig, cmd: &SaleCommand) -> Result<()> {
    let read_client = Arc::new(provider(cfg)?);

    match cmd {
        SaleCommand::Status(target) => show_status(cfg, read_client, target).await?,
        SaleCommand::Claimable { target, user } => {
            show_claimable(cfg, read_client, target, *user).await?
        }
        SaleCommand::Refundable { target, user } => {
            show_refundable(cfg, read_client, target, *user).await?
        }
        SaleCommand::Commitment { target, user } => {
            show_commitment(cfg, read_client, target, *user).await?
        }
        SaleCommand::ApproveCollateral {
            target,
            amount,
            raw,
        } => approve_collateral(cli, cfg, target, amount, *raw).await?,
        SaleCommand::Commit {
            target,
            amount,
            raw,
        } => commit(cli, cfg, target, amount, *raw).await?,
        SaleCommand::Finalize(target) => {
            run_sale_tx(cli, cfg, target, |sale| sale.finalize()).await?
        }
        SaleCommand::Claim(target) => run_sale_tx(cli, cfg, target, |sale| sale.claim()).await?,
        SaleCommand::Refund(target) => run_sale_tx(cli, cfg, target, |sale| sale.refund()).await?,
        SaleCommand::EmergencyRefund(target) => {
            run_sale_tx(cli, cfg, target, |sale| sale.emergency_refund()).await?
        }
    }

    Ok(())
}

async fn show_status(
    cfg: &RuntimeConfig,
    read_client: Arc<Provider<Http>>,
    target: &SaleTarget,
) -> Result<()> {
    let sale_address = resolve_sale_address(read_client.clone(), cfg.factory, target).await?;
    let sale = SaleContract::new(sale_address, read_client.clone());
    let collateral = read_sale_collateral(read_client.clone(), sale_address).await?;
    let (decimals, symbol) = token_meta(read_client.clone(), collateral).await?;
    let (total, accepted, finalized, failed) = sale.get_status().call().await?;

    println!("sale: {:?}", sale_address);
    if let Ok(project_id) = read_sale_project_id(read_client.clone(), sale_address).await {
        println!("project_id: {}", project_id);
        if let Ok(approved) = AgentRaiseFactoryContract::new(cfg.factory, read_client.clone())
            .is_project_approved(project_id)
            .call()
            .await
        {
            println!("project_approved: {}", approved);
        }
    }
    println!("collateral: {:?} ({})", collateral, symbol);
    println!("start_time: {}", sale.start_time().call().await?);
    println!("end_time: {}", sale.end_time().call().await?);
    println!("is_active: {}", sale.is_active().call().await?);
    println!(
        "time_remaining_seconds: {}",
        sale.time_remaining().call().await?
    );
    println!(
        "total_committed: {} ({})",
        format_token_amount(total, decimals),
        total
    );
    println!(
        "accepted_amount: {} ({})",
        format_token_amount(accepted, decimals),
        accepted
    );
    println!("finalized: {}", finalized);
    println!("failed: {}", failed);
    println!("share_token: {:?}", sale.token().call().await?);
    Ok(())
}

async fn show_claimable(
    cfg: &RuntimeConfig,
    read_client: Arc<Provider<Http>>,
    target: &SaleTarget,
    user: Address,
) -> Result<()> {
    let sale_address = resolve_sale_address(read_client.clone(), cfg.factory, target).await?;
    let sale = SaleContract::new(sale_address, read_client.clone());
    let collateral = read_sale_collateral(read_client.clone(), sale_address).await?;
    let (decimals, symbol) = token_meta(read_client, collateral).await?;
    let (payout, refund) = sale.get_claimable(user).call().await?;

    println!("sale: {:?}", sale_address);
    println!("user: {:?}", user);
    println!("token: {}", symbol);
    println!(
        "payout: {} ({})",
        format_token_amount(payout, decimals),
        payout
    );
    println!(
        "refund: {} ({})",
        format_token_amount(refund, decimals),
        refund
    );
    Ok(())
}

async fn show_refundable(
    cfg: &RuntimeConfig,
    read_client: Arc<Provider<Http>>,
    target: &SaleTarget,
    user: Address,
) -> Result<()> {
    let sale_address = resolve_sale_address(read_client.clone(), cfg.factory, target).await?;
    let sale = SaleContract::new(sale_address, read_client.clone());
    let collateral = read_sale_collateral(read_client.clone(), sale_address).await?;
    let (decimals, symbol) = token_meta(read_client, collateral).await?;
    let refundable = sale.get_refundable(user).call().await?;

    println!("sale: {:?}", sale_address);
    println!("user: {:?}", user);
    println!("token: {}", symbol);
    println!(
        "refundable: {} ({})",
        format_token_amount(refundable, decimals),
        refundable
    );
    Ok(())
}

async fn show_commitment(
    cfg: &RuntimeConfig,
    read_client: Arc<Provider<Http>>,
    target: &SaleTarget,
    user: Address,
) -> Result<()> {
    let sale_address = resolve_sale_address(read_client.clone(), cfg.factory, target).await?;
    let sale = SaleContract::new(sale_address, read_client.clone());
    let collateral = read_sale_collateral(read_client.clone(), sale_address).await?;
    let (decimals, symbol) = token_meta(read_client, collateral).await?;
    let committed = sale.commitments(user).call().await?;

    println!("sale: {:?}", sale_address);
    println!("user: {:?}", user);
    println!("token: {}", symbol);
    println!(
        "committed: {} ({})",
        format_token_amount(committed, decimals),
        committed
    );
    Ok(())
}

async fn approve_collateral(
    cli: &Cli,
    cfg: &RuntimeConfig,
    target: &SaleTarget,
    amount: &str,
    raw: bool,
) -> Result<()> {
    let write_client = signer_client(cli, cfg)?;
    let sale_address = resolve_sale_address(write_client.clone(), cfg.factory, target).await?;
    let collateral = read_sale_collateral(write_client.clone(), sale_address).await?;
    let token = Erc20Metadata::new(collateral, write_client.clone());
    let (decimals, symbol) = token_meta(write_client.clone(), collateral).await?;
    let amount_raw = parse_token_amount(amount, decimals, raw)?;

    let receipt = send_and_wait(token.approve(sale_address, amount_raw)).await?;
    println!("owner: {:?}", write_client.address());
    println!(
        "approved: {} {} ({})",
        format_token_amount(amount_raw, decimals),
        symbol,
        amount_raw
    );
    print_receipt(&receipt);
    Ok(())
}

async fn commit(
    cli: &Cli,
    cfg: &RuntimeConfig,
    target: &SaleTarget,
    amount: &str,
    raw: bool,
) -> Result<()> {
    let write_client = signer_client(cli, cfg)?;
    let sale_address = resolve_sale_address(write_client.clone(), cfg.factory, target).await?;
    let collateral = read_sale_collateral(write_client.clone(), sale_address).await?;
    let sale = SaleContract::new(sale_address, write_client.clone());
    let token = Erc20Metadata::new(collateral, write_client.clone());
    let (decimals, symbol) = token_meta(write_client.clone(), collateral).await?;
    let amount_raw = parse_token_amount(amount, decimals, raw)?;
    let sender = write_client.address();

    let allowance = token.allowance(sender, sale_address).call().await?;
    if allowance < amount_raw {
        bail!(
            "insufficient allowance: allowance={} required={}. Run `sale approve-collateral` first.",
            format_token_amount(allowance, decimals),
            format_token_amount(amount_raw, decimals)
        );
    }

    let balance = token.balance_of(sender).call().await?;
    if balance < amount_raw {
        bail!(
            "insufficient balance: balance={} required={}",
            format_token_amount(balance, decimals),
            format_token_amount(amount_raw, decimals)
        );
    }

    let receipt = send_and_wait(sale.commit(amount_raw)).await?;
    println!("sender: {:?}", sender);
    println!(
        "committed: {} {} ({})",
        format_token_amount(amount_raw, decimals),
        symbol,
        amount_raw
    );
    print_receipt(&receipt);
    Ok(())
}

async fn run_sale_tx<F, D>(
    cli: &Cli,
    cfg: &RuntimeConfig,
    target: &SaleTarget,
    build: F,
) -> Result<()>
where
    F: FnOnce(
        SaleContract<crate::chain::contracts::SignerClient>,
    )
        -> ethers::contract::builders::ContractCall<crate::chain::contracts::SignerClient, D>,
    D: ethers::abi::Detokenize,
{
    let write_client = signer_client(cli, cfg)?;
    let sale_address = resolve_sale_address(write_client.clone(), cfg.factory, target).await?;
    let receipt = send_and_wait(build(SaleContract::new(sale_address, write_client))).await?;
    print_receipt(&receipt);
    Ok(())
}

async fn resolve_sale_address<M: Middleware + 'static>(
    client: Arc<M>,
    factory: Address,
    target: &SaleTarget,
) -> Result<Address> {
    if let Some(address) = target.sale {
        return Ok(address);
    }
    if let Some(project_id) = target.project_id {
        return Ok(read_factory_project(client, factory, project_id)
            .await?
            .sale);
    }
    bail!("set --sale or --project-id")
}

fn parse_token_amount(amount: &str, decimals: u8, raw: bool) -> Result<U256> {
    if raw {
        parse_amount_raw(amount)
    } else {
        parse_amount_units(amount, decimals)
    }
}

File: src/commands/sale.ts
import { type Contract, type ContractRunner, type ContractTransactionResponse } from "ethers";
import type { GlobalOptions, SaleCommand, SaleTarget } from "../cli/types.ts";
import type { RuntimeConfig } from "../config/runtime.ts";
import {
  createErc20Contract,
  createFactoryContract,
  createReadClient,
  createSaleContract,
  createShareTokenContract,
  createWriteClient,
  ensureSaleTarget,
  normalizeAddress,
} from "../chain/client.ts";
import {
  readFactoryProject,
  readSaleCollateral,
  readSaleProjectId,
  readTokenMetadata,
} from "../chain/contracts.ts";
import { printReceipt } from "../lib/output.ts";
import { formatTokenAmount, parseAmountRaw, parseAmountUnits, sendAndWait } from "../lib/utils.ts";

export async function runSale(
  global: GlobalOptions,
  config: RuntimeConfig,
  command: SaleCommand,
): Promise<void> {
  const readClient = createReadClient(config);

  switch (command.kind) {
    case "status":
      await showStatus(config, readClient, command.target);
      return;
    case "claimable":
      await showClaimable(config, readClient, command.target, command.user);
      return;
    case "refundable":
      await showRefundable(config, readClient, command.target, command.user);
      return;
    case "commitment":
      await showCommitment(config, readClient, command.target, command.user);
      return;
    case "approve-collateral":
      await approveCollateral(global, config, command.target, command.amount, command.raw);
      return;
    case "commit":
      await commit(global, config, command.target, command.amount, command.raw);
      return;
    case "finalize":
      await runSaleTransaction(global, config, command.target, (sale) => sale.finalize());
      return;
    case "claim":
      await runSaleTransaction(global, config, command.target, (sale) => sale.claim());
      return;
    case "refund":
      await runSaleTransaction(global, config, command.target, (sale) => sale.refund());
      return;
    case "emergency-refund":
      await runSaleTransaction(global, config, command.target, (sale) => sale.emergencyRefund());
      return;
  }
}

async function showStatus(
  config: RuntimeConfig,
  readClient: ReturnType<typeof createReadClient>,
  target: SaleTarget,
): Promise<void> {
  const saleAddress = await resolveSaleAddress(config, readClient, target);
  const sale = createSaleContract(saleAddress, readClient);
  const [totalCommitted, acceptedAmount, finalized, failed] = (await sale.getStatus()) as readonly [
    bigint,
    bigint,
    boolean,
    boolean,
  ];
  const collateral = await readSaleCollateral(saleAddress, readClient);
  const { decimals, symbol } = await readTokenMetadata(collateral, readClient);

  console.log(`sale: ${saleAddress}`);
  try {
    const projectId = await readSaleProjectId(saleAddress, readClient);
    console.log(`project_id: ${projectId}`);
    console.log(
      `project_approved: ${Boolean(
        await createFactoryContract(config.factory, readClient).isProjectApproved(projectId),
      )}`,
    );
  } catch {}
  console.log(`collateral: ${collateral} (${symbol})`);
  console.log(`start_time: ${String(await sale.startTime())}`);
  console.log(`end_time: ${String(await sale.endTime())}`);
  console.log(`is_active: ${Boolean(await sale.isActive())}`);
  console.log(`time_remaining_seconds: ${String(await sale.timeRemaining())}`);
  console.log(
    `total_committed: ${formatTokenAmount(totalCommitted, decimals)} (${totalCommitted})`,
  );
  console.log(
    `accepted_amount: ${formatTokenAmount(acceptedAmount, decimals)} (${acceptedAmount})`,
  );
  console.log(`finalized: ${finalized}`);
  console.log(`failed: ${failed}`);
  const shareTokenAddress = String(await sale.token());
  console.log(`share_token: ${shareTokenAddress}`);
  if (shareTokenAddress !== "0x0000000000000000000000000000000000000000") {
    const shareToken = createShareTokenContract(shareTokenAddress, readClient);
    try {
      const poolAssets = (await shareToken.totalAssets()) as bigint;
      console.log(`fund_term_end: ${String(await shareToken.LOCKUP_END_TIME())}`);
      console.log(`settled: ${Boolean(await shareToken.settled())}`);
      console.log(`redeem_pool: ${formatTokenAmount(poolAssets, decimals)} (${poolAssets})`);
    } catch {}
  }
}

async function showClaimable(
  config: RuntimeConfig,
  readClient: ReturnType<typeof createReadClient>,
  target: SaleTarget,
  user: string,
): Promise<void> {
  const saleAddress = await resolveSaleAddress(config, readClient, target);
  const sale = createSaleContract(saleAddress, readClient);
  const normalizedUser = normalizeAddress(user, "user");
  const collateral = await readSaleCollateral(saleAddress, readClient);
  const { decimals: collateralDecimals, symbol: collateralSymbol } =
    await readTokenMetadata(collateral, readClient);
  const [payout, refund] = (await sale.getClaimable(normalizedUser)) as readonly [bigint, bigint];
  const shareTokenAddress = String(await sale.token());
  const hasShareToken = shareTokenAddress !== "0x0000000000000000000000000000000000000000";
  const { decimals: shareDecimals, symbol: shareSymbol } = hasShareToken
    ? await readTokenMetadata(shareTokenAddress, readClient)
    : { decimals: 18, symbol: "SHARE" };

  console.log(`sale: ${saleAddress}`);
  console.log(`user: ${normalizedUser}`);
  console.log(`share_token: ${shareSymbol}`);
  console.log(
    `claimable_shares: ${formatTokenAmount(payout, shareDecimals)} ${shareSymbol} (${payout})`,
  );
  console.log(
    `overflow_refund: ${formatTokenAmount(refund, collateralDecimals)} ${collateralSymbol} (${refund})`,
  );
}

async function showRefundable(
  config: RuntimeConfig,
  readClient: ReturnType<typeof createReadClient>,
  target: SaleTarget,
  user: string,
): Promise<void> {
  const saleAddress = await resolveSaleAddress(config, readClient, target);
  const sale = createSaleContract(saleAddress, readClient);
  const normalizedUser = normalizeAddress(user, "user");
  const collateral = await readSaleCollateral(saleAddress, readClient);
  const { decimals, symbol } = await readTokenMetadata(collateral, readClient);
  const refundable = (await sale.getRefundable(normalizedUser)) as bigint;

  console.log(`sale: ${saleAddress}`);
  console.log(`user: ${normalizedUser}`);
  console.log(`token: ${symbol}`);
  console.log(`refundable: ${formatTokenAmount(refundable, decimals)} (${refundable})`);
}

async function showCommitment(
  config: RuntimeConfig,
  readClient: ReturnType<typeof createReadClient>,
  target: SaleTarget,
  user: string,
): Promise<void> {
  const saleAddress = await resolveSaleAddress(config, readClient, target);
  const sale = createSaleContract(saleAddress, readClient);
  const normalizedUser = normalizeAddress(user, "user");
  const collateral = await readSaleCollateral(saleAddress, readClient);
  const { decimals, symbol } = await readTokenMetadata(collateral, readClient);
  const committed = (await sale.commitments(normalizedUser)) as bigint;

  console.log(`sale: ${saleAddress}`);
  console.log(`user: ${normalizedUser}`);
  console.log(`token: ${symbol}`);
  console.log(`committed: ${formatTokenAmount(committed, decimals)} (${committed})`);
}

async function approveCollateral(
  global: GlobalOptions,
  config: RuntimeConfig,
  target: SaleTarget,
  amount: string,
  raw: boolean,
): Promise<void> {
  const writeClient = createWriteClient(global, config);
  const saleAddress = await resolveSaleAddress(config, writeClient, target);
  const collateral = await readSaleCollateral(saleAddress, writeClient);
  const token = createErc20Contract(collateral, writeClient);
  const { decimals, symbol } = await readTokenMetadata(collateral, writeClient);
  const value = raw ? parseAmountRaw(amount) : parseAmountUnits(amount, decimals);

  const receipt = await sendAndWait(token.approve(saleAddress, value));
  console.log(`owner: ${writeClient.address}`);
  console.log(`approved: ${formatTokenAmount(value, decimals)} ${symbol} (${value})`);
  printReceipt(receipt);
}

async function commit(
  global: GlobalOptions,
  config: RuntimeConfig,
  target: SaleTarget,
  amount: string,
  raw: boolean,
): Promise<void> {
  const writeClient = createWriteClient(global, config);
  const saleAddress = await resolveSaleAddress(config, writeClient, target);
  const sale = createSaleContract(saleAddress, writeClient);
  const collateral = await readSaleCollateral(saleAddress, writeClient);
  const token = createErc20Contract(collateral, writeClient);
  const { decimals, symbol } = await readTokenMetadata(collateral, writeClient);
  const value = raw ? parseAmountRaw(amount) : parseAmountUnits(amount, decimals);
  const owner = writeClient.address;
  const allowance = (await token.allowance(owner, saleAddress)) as bigint;
  const balance = (await token.balanceOf(owner)) as bigint;

  if (allowance < value) {
    throw new Error(
      `insufficient allowance: allowance=${formatTokenAmount(allowance, decimals)} required=${formatTokenAmount(value, decimals)}`,
    );
  }
  if (balance < value) {
    throw new Error(
      `insufficient balance: balance=${formatTokenAmount(balance, decimals)} required=${formatTokenAmount(value, decimals)}`,
    );
  }

  const receipt = await sendAndWait(sale.commit(value));
  console.log(`sender: ${owner}`);
  console.log(`committed: ${formatTokenAmount(value, decimals)} ${symbol} (${value})`);
  printReceipt(receipt);
}

async function runSaleTransaction(
  global: GlobalOptions,
  config: RuntimeConfig,
  target: SaleTarget,
  build: (sale: Contract) => Promise<ContractTransactionResponse>,
): Promise<void> {
  const writeClient = createWriteClient(global, config);
  const saleAddress = await resolveSaleAddress(config, writeClient, target);
  const sale = createSaleContract(saleAddress, writeClient);
  printReceipt(await sendAndWait(build(sale)));
}

async function resolveSaleAddress(
  config: RuntimeConfig,
  runner: ContractRunner,
  target: SaleTarget,
): Promise<string> {
  const normalizedTarget = ensureSaleTarget(target);
  if (normalizedTarget.sale) {
    return normalizeAddress(normalizedTarget.sale, "sale");
  }
  return (await readFactoryProject(config.factory, runner, normalizedTarget.projectId!)).sale;
}
